src/mono/mono/metadata/gc.c

```diff
@@ -892,6 +896,12 @@ finalizer_thread (gpointer unused)
 		}
 		wait = TRUE;
 
+		/* Just in case we've received a SIGTERM */
+		if (mono_term_signaled) {
```

Comment:
Since this is updated from a different thread without any locking, maybe we would need a read memory barrier to make sure we see updated value, especially when wait == FALSE,. There might be one hidden in the mono_coop_sem_timedwait, looks like at least sem_trywait seems to issue memory barriers, and other implementations probably do as well (like WaitForSingleObjectEx), so maybe not needed in the end, but we depend on implementation details I guess we can leave it as is, just wanted to make a comment/note around potential issue.

Reply:
I think that at one point, this PR had `mono_term_signaled` as `volatile`.

Maybe that (or something else) is needed to ensure these lines don't reorder, causing the wrong exit code to be picked up.

This is how the exit code is set:

```
mono_environment_exitcode_set(128+SIGTERM);	/* Set default exit code */
mono_term_signaled = TRUE;
```

Reply:
We can always do acquire/release semantics on mono_term_signaled and we would make sure this is not reordered (volatile will only prevent compiler to reorder load/stores).